File: src/lib/services/blog-service.ts
import { v4 as uuidv4 } from 'uuid'
import { getCollection, COLLECTIONS } from '@/lib/db/mongodb'
import type { BlogPost, BlogCategory, ServiceStatus, BlogSEO, AIMetadata } from '@/lib/models/types'

function generateSlug(title: string): string {
  return title
    .toLowerCase()
    .replace(/[^\w\sğüşıöç-]/g, '')
    .replace(/\s+/g, '-')
    .replace(/ğ/g, 'g')
    .replace(/ü/g, 'u')
    .replace(/ş/g, 's')
    .replace(/ı/g, 'i')
    .replace(/ö/g, 'o')
    .replace(/ç/g, 'c')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '')
}

export class BlogService {
  // Get all blog posts (admin)
  static async list(
    page: number = 1,
    limit: number = 20,
    filters?: { status?: ServiceStatus; archived?: boolean; search?: string; category?: string }
  ): Promise<{ posts: BlogPost[]; total: number }> {
    const postsCollection = await getCollection<BlogPost>(COLLECTIONS.BLOG_POSTS)
    
    const filter: Record<string, unknown> = {}
    
    if (filters?.archived) {
      filter.archived_at = { $ne: null }
    } else {
      filter.archived_at = null
    }
    
    if (filters?.status) {
      filter.status = filters.status
    }
    
    if (filters?.category) {
      filter.category = filters.category
    }
    
    if (filters?.search) {
      filter.$or = [
        { title: { $regex: filters.search, $options: 'i' } },
        { excerpt: { $regex: filters.search, $options: 'i' } }
      ]
    }
    
    const total = await postsCollection.countDocuments(filter)
    const posts = await postsCollection
      .find(filter)
      .sort({ created_at: -1 })
      .skip((page - 1) * limit)
      .limit(limit)
      .toArray()
    
    return { posts, total }
  }
  
  // Get published blog posts (public)
  static async getPublished(): Promise<{ posts: BlogPost[]; total: number }> {
    const postsCollection = await getCollection<BlogPost>(COLLECTIONS.BLOG_POSTS)
    
    const filter = {
      status: 'published',
      archived_at: null
    }
    
    const total = await postsCollection.countDocuments(filter)
    const posts = await postsCollection
      .find(filter)
      .sort({ published_at: -1 })
      .toArray()
    
    return { posts, total }
  }
  
  // Get by ID
  static async getById(id: string): Promise<BlogPost | null> {
    const postsCollection = await getCollection<BlogPost>(COLLECTIONS.BLOG_POSTS)
    return postsCollection.findOne({ id })
  }
  
  // Get by slug (public)
  static async getPublishedBySlug(slug: string): Promise<BlogPost | null> {
    const postsCollection = await getCollection<BlogPost>(COLLECTIONS.BLOG_POSTS)
    return postsCollection.findOne({
      slug,
      status: 'published',
      archived_at: null
    })
  }
  
  // Create blog post
  static async create(data: {
    title: string
    excerpt?: string
    cover_image_url?: string
    cover_image_alt?: string
    content?: string
    category?: string
    tags?: string[]
    seo?: BlogSEO
    status?: ServiceStatus
    ai_generated?: boolean
    ai_metadata?: AIMetadata
    created_by?: string
  }): Promise<BlogPost> {
    const postsCollection = await getCollection<BlogPost>(COLLECTIONS.BLOG_POSTS)
    
    // Generate unique slug
    let slug = generateSlug(data.title)
    let counter = 1
    while (await postsCollection.findOne({ slug })) {
      slug = `${generateSlug(data.title)}-${counter}`
      counter++
    }
    
    const now = new Date()
    const post: BlogPost = {
      id: uuidv4(),
      title: data.title,
      slug,
      excerpt: data.excerpt || null,
      cover_image_url: data.cover_image_url || null,
      cover_image_alt: data.cover_image_alt || null,
      content: data.content || null,
      category: data.category || null,
      tags: data.tags || [],
      seo: data.seo || null,
      status: data.status || 'draft',
      published_at: data.status === 'published' ? now : null,
      ai_generated: data.ai_generated || false,
      ai_metadata: data.ai_metadata || null,
      view_count: 0,
      archived_at: null,
      created_at: now,
      updated_at: now,
      created_by: data.created_by || null,
      updated_by: null
    }
    
    await postsCollection.insertOne(post)
    return post
  }
  
  // Update blog post
  static async update(id: string, data: Partial<BlogPost>, updatedBy?: string): Promise<BlogPost | null> {
    const postsCollection = await getCollection<BlogPost>(COLLECTIONS.BLOG_POSTS)
    
    const current = await this.getById(id)
    if (!current) return null
    
    const updateData: Record<string, unknown> = {
      ...data,
      updated_at: new Date(),
      updated_by: updatedBy || null
    }
    
    // Set published_at if publishing
    if (data.status === 'published' && current.status !== 'published') {
      updateData.published_at = new Date()
    }
    
    delete updateData.id
    delete updateData.created_at
    delete updateData.created_by
    
    await postsCollection.updateOne(
      { id },
      { $set: updateData }
    )
    
    return this.getById(id)
  }
  
  // Update status
  static async updateStatus(id: string, status: ServiceStatus): Promise<BlogPost | null> {
    return this.update(id, { status })
  }
  
  // Increment view count
  static async incrementViewCount(slug: string, clientIp: string): Promise<number> {
    const postsCollection = await getCollection<BlogPost>(COLLECTIONS.BLOG_POSTS)
    
    // Simple increment (in production, implement deduplication with Redis)
    const result = await postsCollection.findOneAndUpdate(
      { slug },
      { $inc: { view_count: 1 } },
      { returnDocument: 'after' }
    )
    
    return result?.view_count || 0
  }
  
  // Archive
  static async archive(id: string): Promise<boolean> {
    const postsCollection = await getCollection<BlogPost>(COLLECTIONS.BLOG_POSTS)
    
    const result = await postsCollection.updateOne(
      { id },
      { $set: { archived_at: new Date(), status: 'archived' as ServiceStatus } }
    )
    
    return result.modifiedCount > 0
  }
  
  // Restore
  static async restore(id: string): Promise<BlogPost | null> {
    const postsCollection = await getCollection<BlogPost>(COLLECTIONS.BLOG_POSTS)
    
    await postsCollection.updateOne(
      { id },
      { $set: { archived_at: null, status: 'draft' as ServiceStatus } }
    )
    
    return this.getById(id)
  }
}

// Blog Categories Service
export class BlogCategoryService {
  static async list(): Promise<BlogCategory[]> {
    const categoriesCollection = await getCollection<BlogCategory>(COLLECTIONS.BLOG_CATEGORIES)
    return categoriesCollection.find({}).sort({ sort_order: 1 }).toArray()
  }
  
  static async getById(id: string): Promise<BlogCategory | null> {
    const categoriesCollection = await getCollection<BlogCategory>(COLLECTIONS.BLOG_CATEGORIES)
    return categoriesCollection.findOne({ id })
  }
  
  static async create(data: {
    name: string
    description?: string
    color?: string
    icon?: string
  }): Promise<BlogCategory> {
    const categoriesCollection = await getCollection<BlogCategory>(COLLECTIONS.BLOG_CATEGORIES)
    
    const slug = generateSlug(data.name)
    
    const maxSort = await categoriesCollection
      .find({})
      .sort({ sort_order: -1 })
      .limit(1)
      .toArray()
    const sortOrder = maxSort.length > 0 ? maxSort[0].sort_order + 1 : 0
    
    const now = new Date()
    const category: BlogCategory = {
      id: uuidv4(),
      name: data.name,
      slug,
      description: data.description || null,
      color: data.color || null,
      icon: data.icon || null,
      sort_order: sortOrder,
      is_active: true,
      post_count: 0,
      created_at: now,
      updated_at: now
    }
    
    await categoriesCollection.insertOne(category)
    return category
  }
  
  static async update(id: string, data: Partial<BlogCategory>): Promise<BlogCategory | null> {
    const categoriesCollection = await getCollection<BlogCategory>(COLLECTIONS.BLOG_CATEGORIES)
    
    const updateData: Record<string, unknown> = {
      ...data,
      updated_at: new Date()
    }
    
    delete updateData.id
    delete updateData.created_at
    
    await categoriesCollection.updateOne(
      { id },
      { $set: updateData }
    )
    
    return this.getById(id)
  }
  
  static async delete(id: string): Promise<boolean> {
    const categoriesCollection = await getCollection<BlogCategory>(COLLECTIONS.BLOG_CATEGORIES)
    const result = await categoriesCollection.deleteOne({ id })
    return result.deletedCount > 0
  }
}

File: backend/services/blog_service.py
from typing import Optional, List, Dict, Any, Tuple
from datetime import datetime, timezone
from db.mongodb import get_database
from models.blog import BlogPost, BlogStatus, BlogSEO
import re
import logging

logger = logging.getLogger(__name__)

def generate_slug(title: str) -> str:
    """Generate URL-friendly slug from title"""
    slug = title.lower()
    # Turkish character replacements
    replacements = {
        'ı': 'i', 'ğ': 'g', 'ü': 'u', 'ş': 's', 'ö': 'o', 'ç': 'c',
        'İ': 'i', 'Ğ': 'g', 'Ü': 'u', 'Ş': 's', 'Ö': 'o', 'Ç': 'c'
    }
    for tr_char, en_char in replacements.items():
        slug = slug.replace(tr_char, en_char)
    slug = re.sub(r'[^a-z0-9\s-]', '', slug)
    slug = re.sub(r'[\s_]+', '-', slug)
    slug = re.sub(r'-+', '-', slug)
    return slug.strip('-')

class BlogService:
    COLLECTION = "blog_posts"
    
    @staticmethod
    def _deserialize_post(doc: dict) -> dict:
        """Convert stored document to BlogPost-compatible dict"""
        if doc.get('seo') and isinstance(doc['seo'], dict):
            doc['seo'] = BlogSEO(**doc['seo'])
        if doc.get('status'):
            doc['status'] = BlogStatus(doc['status'])
        for date_field in ['created_at', 'updated_at', 'published_at', 'archived_at']:
            if doc.get(date_field) and isinstance(doc[date_field], str):
                doc[date_field] = datetime.fromisoformat(doc[date_field].replace('Z', '+00:00'))
        return doc
    
    @staticmethod
    async def create(
        title: str,
        created_by: Optional[str] = None,
        author_id: Optional[str] = None,
        slug: Optional[str] = None,
        meta_title: Optional[str] = None,
        meta_description: Optional[str] = None,
        ai_generated: bool = False,
        ai_metadata: Optional[dict] = None,
        **kwargs
    ) -> BlogPost:
        """Create a new blog post"""
        db = get_database()
        
        if not slug:
            slug = generate_slug(title)
        
        # Ensure unique slug
        base_slug = slug
        counter = 1
        while await db[BlogService.COLLECTION].find_one({"slug": slug}):
            slug = f"{base_slug}-{counter}"
            counter += 1
        
        # Build SEO object if meta fields provided
        seo = None
        if meta_title or meta_description:
            seo = BlogSEO(
                meta_title=meta_title,
                meta_description=meta_description
            )
        # If seo was passed in kwargs, use it (from API) or build from meta fields
        if 'seo' in kwargs:
            seo_data = kwargs.pop('seo')
            if seo_data and isinstance(seo_data, dict):
                seo = BlogSEO(**seo_data)
            elif seo_data and isinstance(seo_data, BlogSEO):
                seo = seo_data
        
        post = BlogPost(
            title=title,
            slug=slug,
            created_by=created_by or author_id,
            seo=seo,
            ai_generated=ai_generated,
            ai_metadata=ai_metadata,
            **kwargs
        )
        
        doc = post.to_dict()
        doc["_id"] = post.id
        
        await db[BlogService.COLLECTION].insert_one(doc)
        logger.info(f"Created blog post: {post.id} - {post.title}")
        
        return post
    
    @staticmethod
    async def get_by_id(post_id: str) -> Optional[BlogPost]:
        """Get blog post by ID"""
        db = get_database()
        doc = await db[BlogService.COLLECTION].find_one({"id": post_id}, {"_id": 0})
        if not doc:
            return None
        doc = BlogService._deserialize_post(doc)
        return BlogPost(**doc)
    
    @staticmethod
    async def get_by_slug(slug: str) -> Optional[BlogPost]:
        """Get blog post by slug"""
        db = get_database()
        doc = await db[BlogService.COLLECTION].find_one({"slug": slug}, {"_id": 0})
        if not doc:
            return None
        doc = BlogService._deserialize_post(doc)
        return BlogPost(**doc)
    
    @staticmethod
    async def get_list(
        page: int = 1,
        page_size: int = 20,
        status: Optional[BlogStatus] = None,
        archived: bool = False,
        search: Optional[str] = None,
        category: Optional[str] = None
    ) -> Tuple[List[BlogPost], int]:
        """Get paginated list of blog posts for admin"""
        db = get_database()
        
        query: Dict[str, Any] = {}
        
        if archived:
            query["archived_at"] = {"$ne": None}
        else:
            query["archived_at"] = None
        
        if status:
            query["status"] = status.value
        
        if category:
            query["category"] = category
        
        if search:
            query["$or"] = [
                {"title": {"$regex": search, "$options": "i"}},
                {"excerpt": {"$regex": search, "$options": "i"}}
            ]
        
        total = await db[BlogService.COLLECTION].count_documents(query)
        
        skip = (page - 1) * page_size
        cursor = db[BlogService.COLLECTION].find(query, {"_id": 0}).skip(skip).limit(page_size).sort("created_at", -1)
        docs = await cursor.to_list(page_size)
        
        posts = []
        for doc in docs:
            doc = BlogService._deserialize_post(doc)
            posts.append(BlogPost(**doc))
        
        return posts, total
    
    @staticmethod
    async def get_published(page: int = 1, page_size: int = 10, category: str = None) -> tuple:
        """Get all published, non-archived posts for public with pagination"""
        db = get_database()
        
        query = {
            "status": BlogStatus.PUBLISHED.value,
            "archived_at": None
        }
        
        if category:
            query["category"] = category
        
        skip = (page - 1) * page_size
        total = await db[BlogService.COLLECTION].count_documents(query)
        
        cursor = db[BlogService.COLLECTION].find(query, {"_id": 0}).sort("published_at", -1).skip(skip).limit(page_size)
        docs = await cursor.to_list(length=page_size)
        
        posts = []
        for doc in docs:
            doc = BlogService._deserialize_post(doc)
            posts.append(BlogPost(**doc))
        
        return posts, total
    
    @staticmethod
    async def get_published_by_slug(slug: str) -> Optional[BlogPost]:
        """Get published post by slug for public"""
        db = get_database()
        
        query = {
            "slug": slug,
            "status": BlogStatus.PUBLISHED.value,
            "archived_at": None
        }
        
        doc = await db[BlogService.COLLECTION].find_one(query, {"_id": 0})
        if not doc:
            return None
        doc = BlogService._deserialize_post(doc)
        return BlogPost(**doc)
    
    @staticmethod
    async def update(post_id: str, updated_by: str, **kwargs) -> Optional[BlogPost]:
        """Update blog post fields"""
        db = get_database()
        
        update_data = {
            "updated_at": datetime.now(timezone.utc).isoformat(),
            "updated_by": updated_by
        }
        
        for key, value in kwargs.items():
            if value is not None:
                if key == 'status' and hasattr(value, 'value'):
                    update_data[key] = value.value
                elif key == 'seo' and hasattr(value, 'model_dump'):
                    update_data[key] = value.model_dump()
                elif key == 'seo' and isinstance(value, dict):
                    update_data[key] = value
                else:
                    update_data[key] = value
        
        # Ensure unique slug if updating
        if 'slug' in update_data:
            existing = await db[BlogService.COLLECTION].find_one({
                "slug": update_data['slug'],
                "id": {"$ne": post_id}
            })
            if existing:
                return None
        
        result = await db[BlogService.COLLECTION].update_one(
            {"id": post_id},
            {"$set": update_data}
        )
        
        if result.modified_count == 0:
            return None
        
        return await BlogService.get_by_id(post_id)
    
    @staticmethod
    async def update_status(post_id: str, status: BlogStatus, updated_by: str) -> Optional[BlogPost]:
        """Update post status (publish/unpublish)"""
        update_data = {"status": status}
        if status == BlogStatus.PUBLISHED:
            update_data["published_at"] = datetime.now(timezone.utc)
        return await BlogService.update(post_id, updated_by, **update_data)
    
    @staticmethod
    async def archive(post_id: str, updated_by: str) -> Optional[BlogPost]:
        """Archive a blog post"""
        return await BlogService.update(
            post_id,
            updated_by,
            archived_at=datetime.now(timezone.utc),
            status=BlogStatus.DRAFT
        )
    
    @staticmethod
    async def restore(post_id: str, updated_by: str) -> Optional[BlogPost]:
        """Restore an archived blog post"""
        db = get_database()
        
        result = await db[BlogService.COLLECTION].update_one(
            {"id": post_id},
            {"$set": {
                "archived_at": None,
                "updated_at": datetime.now(timezone.utc).isoformat(),
                "updated_by": updated_by
            }}
        )
        
        if result.modified_count == 0:
            return None
        
        return await BlogService.get_by_id(post_id)


    @staticmethod
    async def increment_view_count(slug: str, client_ip: str) -> int:
        """Increment view count for a blog post with basic IP deduplication"""
        db = get_database()
        
        # Simple deduplication: check if this IP viewed in last 30 minutes
        view_key = f"{slug}:{client_ip}"
        views_collection = db["blog_views"]
        
        # Check for recent view from same IP
        recent_view = await views_collection.find_one({
            "key": view_key,
            "timestamp": {"$gte": datetime.now(timezone.utc).timestamp() - 1800}  # 30 minutes
        })
        
        if not recent_view:
            # Record this view
            await views_collection.update_one(
                {"key": view_key},
                {
                    "$set": {
                        "key": view_key,
                        "timestamp": datetime.now(timezone.utc).timestamp()
                    }
                },
                upsert=True
            )
            
            # Increment view count
            await db[BlogService.COLLECTION].update_one(
                {"slug": slug},
                {"$inc": {"view_count": 1}}
            )
        
        # Return current view count
        post = await db[BlogService.COLLECTION].find_one({"slug": slug}, {"view_count": 1, "_id": 0})
        return post.get("view_count", 0) if post else 0
    
    @staticmethod
    async def get_view_count(slug: str) -> int:
        """Get view count for a blog post"""
        db = get_database()
        post = await db[BlogService.COLLECTION].find_one({"slug": slug}, {"view_count": 1, "_id": 0})
        return post.get("view_count", 0) if post else 0
